agroal-api/src/main/java/io/agroal/api/cache/LocalConnectionCache.java

```diff
@@ -0,0 +1,137 @@
+// Copyright (C) 2017 Red Hat, Inc. and individual contributors as indicated by the @author tags.
+// You may not use this file except in compliance with the Apache License, Version 2.0.
+
+package io.agroal.api.cache;
+
+import java.util.ArrayDeque;
+import java.util.Deque;
+
+/**
+ * Implementations of ConnectionCache that rely on {@link ThreadLocal}.
+ *
+ * @author <a href="lbarreiro@redhat.com">Luis Barreiro</a>
+ */
+public interface LocalConnectionCache {
+
+    /**
+     * A local cache that stores at most a single connection.
+     */
+    static ConnectionCache single() {
+        return new ConnectionCache() {
+
+            @SuppressWarnings( "ThreadLocalNotStaticFinal" )
+            private ThreadLocal<Acquirable> threadLocal;
+
+            { // instance initializer
+                reset();
+            }
+
+            @Override
+            public Acquirable get() {
+                Acquirable acquirable = threadLocal.get();
+                return acquirable != null && acquirable.acquire() ? acquirable : null;
+            }
+
+            @Override
+            public void put(Acquirable acquirable) {
+                if ( acquirable.isAcquirable() ) {
+                    threadLocal.set( acquirable );
+                }
+            }
+
+            @Override
+            public void reset() {
+                threadLocal = new ThreadLocal<>();
+            }
+        };
+    }
+
+    /**
+     * A local cache that stores up to a number of connections.
+     */
+    static ConnectionCache fixed(int size) {
+        return new ConnectionCache() {
+
+            @SuppressWarnings( "ThreadLocalNotStaticFinal" )
+            private ThreadLocal<Acquirable[]> threadLocal;
+
+            { // instance initializer
+                reset();
+            }
+
+            @Override
+            public Acquirable get() {
+                Acquirable[] cacheArray = threadLocal.get();
+                for ( int i = cacheArray.length; i > 0; ) { // iterate from last to first
+                    Acquirable element = cacheArray[--i];
+                    if ( element != null ) {
+                        cacheArray[i] = null;
+                        if ( element.acquire() ) {
+                            return element;
+                        }
+                    }
+                }
+                return null;
+            }
+
+            @Override
+            public void put(Acquirable acquirable) {
+                Acquirable[] cacheArray = threadLocal.get();
+                int last = cacheArray.length - 1;
+                if ( cacheArray[last] != null ) { // replace last
```

Comment:
What is the reasoning behind the logic here? It fills the last spot first, then fills from the front, but aquires from the back?

If you have two items in here this means you get one at the front and one at the back? If there is a reason for this can you add a comment explaining why?

Reply:
the logic is that the latest put has always preference, as it will be likely that can be acquired again. the other slots are remembered in the hope that whoever acquired it had released it between. 

anyway, this implementation is to mimic the existing behavior (for benchmarking purposes) and the plan is to change the default behavior (in favor of one of the other implementations, likely `single()`).

Reply:
changed the insert order to fill from last to first.